### ISSUE TITLE
Setting use_structure_from_insertion_table_in_table_functions seems error-prone with parquet

### ISSUE THREAD

**Issue by al13n321** [ClickHouse/ClickHouse#70956]
`INSERT ... SELECT * FROM file(..., Parquet)`, by default, effectively does column matching by name instead of position. E.g. if the parquet file has columns in different order they'll be reordered, and if it has extra columns the query won't fail. This seems likely to cause confusion in practice because:
 * In ~all other situations columns are matched by position: in standard SQL, when doing INSERT SELECT from a regular table, etc.
 * Small perturbations of the query may flip the behavior back to matching by position. E.g. specifying the list of columns explicitly in SELECT list or in file(...) arguments, or changing `use_structure_from_insertion_table_in_table_functions` setting.
 * Seems likely that we'll accidentally change how `use_structure_from_insertion_table_in_table_functions = 2` works in the code, such that sometimes it'll fall back to position-based matching. (E.g. in https://github.com/ClickHouse/ClickHouse/issues/53157 I was thinking of falling back to schema inference whenever WHERE is present. Then I noticed this whole issue.)

E.g. maybe someone used to do INSERT SELECT from parquet files, then changed the pipeline to use an intermediate temporary table and INSERT SELECT from that, and suddenly it stops working.

This applies to input formats that have column names in the file (e.g. parquet, avro, orc; but not e.g. csv), when the setting `use_structure_from_insertion_table_in_table_functions` is not disabled.

---

Examples of unexpected behavior:

```sql
insert into function file('t.parquet') select 'y' as y, 'x' as x settings engine_file_truncate_on_insert=1;
create table a (x String, y String) engine Memory;
insert into a select * from file('t.parquet');
insert into a select * from file('t.parquet') settings use_structure_from_insertion_table_in_table_functions=0;

:) select * from a

SELECT *
FROM a

Query id: 78680961-01b2-4f51-bdf1-1efd0ec94f0f

   ┌─x─┬─y─┐
1. │ y │ x │                <--- correct
   └───┴───┘
   ┌─x─┬─y─┐
2. │ x │ y │
   └───┴───┘

create table b (z String, w String) engine Memory;
insert into b select * from file('t.parquet');
insert into b select * from file('t.parquet') settings use_structure_from_insertion_table_in_table_functions=0;

:) select * from b

SELECT *
FROM b

Query id: 16593bc8-63b2-4c7d-96db-9780081cdcb5

   ┌─z─┬─w─┐
1. │   │   │
   └───┴───┘
   ┌─z─┬─w─┐
2. │ y │ x │                <--- correct
   └───┴───┘
```

---

A simple fix would be to make `use_structure_from_insertion_table_in_table_functions = 2` not apply to `*`, do schema inference in that case. But the name-based `INSERT SELECT *` seems really useful when ingesting many parquet files whose schemas may be different, may have extra columns, and may change over time. We can keep the behavior of `use_structure_from_insertion_table_in_table_functions = 1` and officially recommend to use it for name-based column matching. But it seems likely that many people are relying on this behavior, and they'll all have to notice the breakage and add the setting, but maybe it's still better than leaving default behavior inconsistent?

---

In other words, this setting changes the meaning of `*` from "all columns in the source table" to "all columns in the target table". Maybe this should be a general feature, with some syntax like `SELECT **`, supported for any INSERT SELECT queries, not specific to input formats? Or is there already a feature like that?

**Comment by al13n321** [comment]
@Avogar , what do you think?

**Comment by Avogar** [comment]
> A simple fix would be to make use_structure_from_insertion_table_in_table_functions = 2 not apply to *, do schema inference in that case. But the name-based INSERT SELECT * seems really useful when ingesting many parquet files whose schemas may be different, may have extra columns, and may change over time. We can keep the behavior of use_structure_from_insertion_table_in_table_functions = 1 and officially recommend to use it for name-based column matching. But it seems likely that many people are relying on this behavior, and they'll all have to notice the breakage and add the setting, but maybe it's still better than leaving default behavior inconsistent?

Yes, I agree that the default behaviour is slightly inconsistent, but it was kind of intentional, `INSERT INTO table SELECT * FROM file/s3/...` is really popular and is used by a lot of users (including our internal teams) to load data into tables. And even with formats with information about column names (like Parquet/ORC/Arrow/Avro/*WithNames/etc) usage of table schema is usually more efficient than using schema inferrence (because in this case we can avoid unnecessary casts during insert select/usage of Nullable columns/better parsing of text formats with proper data types, etc). Also schema inference for some formats takes some time (for Parquet it's more efficient because of metadata, but we still need to read it twice - once during schema inference, second during data reading). But yes, we also have schema inference cache, so we will do it only on first read from the file.

And if we change it, it would be a noticable change for the users, so I am not sure about it. I see that `use_structure_from_insertion_table_in_table_functions=2` doesn't work well when query contains `WHERE/GROUP BY/ORDER BY` and probably we should just fix it.

@alexey-milovidov WDYT?

**Comment by alexey-milovidov** [comment]
I agree with @Avogar - for formats with column names, it makes sense to adapt the order of columns to the INSERT's data structure. We are effectively asking the format to give us the columns needed for INSERT.

Yes, it is true that changing the query slightly, e.g., specifying the columns explicitly, will inhibit this behavior... We need more thought about this.
